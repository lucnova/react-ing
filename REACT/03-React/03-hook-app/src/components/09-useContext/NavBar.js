import React from 'react';
import { Link, NavLink } from 'react-router-dom';

export const NavBar = () => {
	return (
		<nav className="navbar navbar-expand-lg navbar-dark bg-dark px-3">
			<Link to="/" className="navbar-brand">
				UseContext
			</Link>

			<div className="collapse navbar-collapse">
				<ul className="navbar-nav mr-auto">
					<NavLink exact to="/" className="nav-item nav-link" activeClassName="active">
						Home
					</NavLink>
					<NavLink exact to="/about" className="nav-item nav-link" activeClassName="active">
						About
					</NavLink>
					<NavLink exact to="/login" className="nav-item nav-link" activeClassName="active">
						Login
					</NavLink>
				</ul>
			</div>
		</nav>
	);
};
