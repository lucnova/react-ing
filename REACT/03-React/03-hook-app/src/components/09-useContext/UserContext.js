// * Higher Order Component
const { createContext } = require("react");

// * Exportarlo
export const UserContext = createContext(null);