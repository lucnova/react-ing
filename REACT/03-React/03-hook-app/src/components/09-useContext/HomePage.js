import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const HomePage = () => {
	const userContext = useContext(UserContext);
	const { user } = userContext;

	return (
		<>
			<h1>Home Page</h1>
			<hr />

			<pre>{JSON.stringify(user, null, 4)}</pre>
		</>
	);
};
