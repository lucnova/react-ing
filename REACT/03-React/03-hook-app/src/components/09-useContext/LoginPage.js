import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const LoginPage = () => {
	const { setUser } = useContext(UserContext);

	return (
		<>
			<h1>Login Page</h1>
			<hr />

			<button
				className="btn btn-primary"
				onClick={() =>
					setUser({
						id: '123ABC',
						username: 'lucnova',
						email: 'alanugb@gmail.com',
					})
				}
			>
				Login
			</button>

			<button className="btn btn-warning" onClick={() => setUser({})}>
				Logout
			</button>
		</>
	);
};
