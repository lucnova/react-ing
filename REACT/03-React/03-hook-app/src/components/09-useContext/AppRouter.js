import React from 'react';
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom';
import { AboutPage } from './AboutPage';
import { HomePage } from './HomePage';
import { LoginPage } from './LoginPage';
import { NavBar } from './NavBar';

export const AppRouter = () => {
	// * El Switch devuelve lo primero que se indique; si pones Route path='/', no importa si te dirijes a /about;
	//       coinciden el / y es suficiente para el Switch,
	//      Para hacerlo exacto, colocar 'exact' como propiedad.
	return (
		<>
			<Router>
				<div>
					<NavBar />
					<Switch>
						<Route exact path="/home" component={HomePage} />
						<Route exact path="/login" component={LoginPage} />
						<Route exact path="/about" component={AboutPage} />
						<Route exact path="/" component={HomePage} />
					</Switch>
				</div>
			</Router>
		</>
	);
};
