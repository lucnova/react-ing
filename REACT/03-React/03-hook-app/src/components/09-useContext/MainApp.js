import React from 'react';
import { AppRouter } from './AppRouter';
import { UserContext } from './UserContext';

// * Establecer el contexto a todo el componente principal.

export const MainApp = () => {
	const user = {
		id: '12ABC',
		username: 'lucnova',
		email: 'nlnovae@gmail.com',
	};

	return (
		<UserContext.Provider value={user}>
			<AppRouter />
		</UserContext.Provider>
	);
};
