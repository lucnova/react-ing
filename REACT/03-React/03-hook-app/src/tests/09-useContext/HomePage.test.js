import { shallow, mount } from 'enzyme';
import { HomePage } from '../../components/09-useContext/HomePage';
import { UserContext } from '../../components/09-useContext/UserContext';

const user = {
	id: 'TEST',
	username: 'lucnova',
	email: 'alanugb@gmail.com',
};

describe('Testing on HomePage', () => {
	// * 2) Al usar shallow envolviendo al HomePage solo renderiza el primer componente.
	//      Cambiar de shallow -> mount.
	// const wrapper = shallow(
	// 	<UserContext.Provider
	// 		value={{
	// 			user,
	// 		}}
	// 	>
	// 		<HomePage />
	// 	</UserContext.Provider>
	// );
	const wrapper = mount(
		<UserContext.Provider
			value={{
				user,
			}}
		>
			<HomePage />
		</UserContext.Provider>
	);

	test('should show component as snapshot', () => {
		// * 1) Si esto se ejecuta as√≠ sin mas, entonces da error: falta contexto en la funcion useContext
		//      Esto es porque estas pruebas son ejecutadas de manera aislada.
		//      Se resuelve envolviendo el HomePage con un UserContext como en el inicio (2).
		expect(wrapper).toMatchSnapshot();
	});
});
