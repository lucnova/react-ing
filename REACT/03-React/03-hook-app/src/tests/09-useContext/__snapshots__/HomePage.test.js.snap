// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on HomePage should show component as snapshot 1`] = `
Array [
  <h1>
    Home Page
  </h1>,
  <hr />,
  <pre>
    {
    "id": "TEST",
    "username": "lucnova",
    "email": "alanugb@gmail.com"
}
  </pre>,
]
`;
